### DISCUSSION TITLE
Test and Production mode

### DISCUSSION THREAD

**Discussion by zeuchste** [nebulastream/nebulastream#13] [Ideas]
As wie have many things where wie discuss in they should be in the Master or Not, one other idea that they also do in other Systems Like rising Wave is that they start the System in Developer Mode and the have a Switch to move it to Production Mode. In this setup we could do many things Like deactivating internal sources @alepping or switching to a high Performance hash join @keyseven123 etc. We could have a Look how they did it as they are Open source

**Comment by PhilippGrulich** [comment]
But you still have the code complexity which you have to maintain if you change Apis.

**Comment by ls-1801** [comment]
@zeuchste, I am not sure; the way you describe it, it sounds like they do the switch at runtime. If that's what you mean, I don't like it because you will always pay for something that you don't need in Production Mode, and vice versa. There will always be tooling that slows down the system, even if disabled.

I would handle all different system configurations at compile time, disabling certain sources within the CMake build.
My problem is that we have a gazillion flags, some of which just don't work. So, we also need to test whatever flags we support in the CI.

We should be stricter with the usage of #ifdefs tho, I would rather have duplicated code instead of very fine granular #ifdefs to disable or enable fields.

**Reply by zeuchste** [reply]
I am not sure if they do it at runtime or compile time. The general idea would be to have two versions one with all features and one with production but you are right it is maybe too much to maintain so it was rather food for thought :)

**Reply by keyseven123** [reply]
Can we then close this discussion here then @zeuchste ?

**Reply by zeuchste** [reply]
yes for now if nobody wants it :)
